ffi/src/lib.rs

```diff
@@ -40,3 +54,741 @@ impl Iterator for EngineIterator {
         }
     }
 }
+
+/// A non-owned slice of a UTF8 string, intended for arg-passing between kernel and engine. The
+/// slice is only valid until the function it was passed into returns, and should not be copied.
+///
+/// # Safety
+///
+/// Intentionally not Copy, Clone, Send, nor Sync.
+///
+/// Whoever instantiates the struct must ensure it does not outlive the data it points to. The
+/// compiler cannot help us here, because raw pointers don't have lifetimes. To reduce the risk of
+/// accidental misuse, it is recommended to only instantiate this struct as a function arg, by
+/// converting a `&str` value `Into<KernelStringSlice>`, so the borrowed reference protects the
+/// function call (callee must not retain any references to the slice after the call returns):
+///
+/// ```
+/// fn wants_slice(slice: KernelStringSlice) { ... }
+/// let msg = String::from(...);
+/// wants_slice(msg.as_ref().into());
+/// ```
+#[repr(C)]
+pub struct KernelStringSlice {
+    ptr: *const c_char,
+    len: usize,
+}
+
+// Intentionally not From, in order to reduce risk of accidental misuse. The main use is for callers
+// to pass e.g. `my_str.as_str().into()` as a function arg.
+#[allow(clippy::from_over_into)]
+impl Into<KernelStringSlice> for &str {
+    fn into(self) -> KernelStringSlice {
+        KernelStringSlice {
+            ptr: self.as_ptr().cast(),
+            len: self.len(),
+        }
+    }
+}
+
+trait TryFromStringSlice: Sized {
+    unsafe fn try_from_slice(slice: KernelStringSlice) -> DeltaResult<Self>;
+}
+
+impl TryFromStringSlice for String {
+    /// Converts a slice back to a string
+    ///
+    /// # Safety
+    ///
+    /// The slice must be a valid (non-null) pointer, and must point to the indicated number of
+    /// valid utf8 bytes.
+    unsafe fn try_from_slice(slice: KernelStringSlice) -> DeltaResult<String> {
+        let slice = unsafe { std::slice::from_raw_parts(slice.ptr.cast(), slice.len) };
+        match std::str::from_utf8(slice) {
+            Ok(s) => Ok(s.to_string()),
+            Err(e) => Err(Error::generic_err(e)),
+        }
```

Comment:
We have a `Utf8Error` `#[from]` variant for our Error, so you should be able to do:
```suggestion
        Ok(std::str::from_utf8(slice)?.to_string())
```

Reply:
```
error[E0277]: `?` couldn't convert the error to `deltakernel::Error`
   --> ffi/src/lib.rs:108:38
    |
108 |         Ok(std::str::from_utf8(slice)?.to_string())
    |                                      ^ the trait `From<std::str::Utf8Error>` is not implemented for `deltakernel::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: the following other types implement trait `From<T>`:
              <deltakernel::Error as From<serde_json::error::Error>>
              <deltakernel::Error as From<arrow_schema::error::ArrowError>>
              <deltakernel::Error as From<url::ParseError>>
              <deltakernel::Error as From<std::io::Error>>
              <deltakernel::Error as From<parquet::errors::ParquetError>>
              <deltakernel::Error as From<FromUtf8Error>>
              <deltakernel::Error as From<std::num::ParseIntError>>
    = note: required for `Result<std::string::String, deltakernel::Error>` to implement `FromResidual<Result<Infallible, std::str::Utf8Error>>`
```

Reply:
just going to leave this as todo for now and we can revisit if we want? (also some discussion of panic instead of errors)

Reply:
hrmm, that's odd, but yeah, i think just `unwrap` and get rid of all of this for now